core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -1,187 +1,337 @@
-    fn update_internal(&self, root: Slot, epoch_slots_list: Vec<EpochSlots>, num_epoch_slots: u64) {
-        // Attach validator's total stake.
-        let epoch_slots_list: Vec<_> = {
-            let validator_stakes = self.validator_stakes.read().unwrap();
-            epoch_slots_list
-                .into_iter()
-                .map(|epoch_slots| {
-                    let stake = match validator_stakes.get(&epoch_slots.from) {
-                        Some(v) => v.total_stake,
-                        None => 0,
-                    };
-                    (epoch_slots, stake)
-                })
-                .collect()
-        };
-        // Discard slots at or before current root or epochs ahead.
-        let slot_range = (root + 1)
-            ..root.saturating_add(
-                num_epoch_slots
-                    .max(DEFAULT_SLOTS_PER_EPOCH)
-                    .saturating_mul(2),
+    /// Advance the cluster_slots ringbuffer, initialize if needed.
+    fn roll_cluster_slots(
+        &self,
+        validator_stakes: &HashMap<Pubkey, u64>,
+        slot_range: &Range<Slot>,
+    ) {
+        let current_slot = self.current_slot.swap(slot_range.start, Ordering::Relaxed);
+        if current_slot >= slot_range.start {
+            return;
+        }
+        let map_capacity = validator_stakes.len();
+        let mut cluster_slots = self.cluster_slots.write().unwrap();
+        self.metric_write_locks.fetch_add(1, Ordering::Relaxed);
+        // zero-initialize a given map
+        let zero_init_map = |map: &mut SlotPubkeys| {
+            map.extend(
+                validator_stakes
+                    .iter()
+                    .map(|(k, _v)| (*k, AtomicU64::new(0))),
             );
-        let slot_nodes_stakes = epoch_slots_list
-            .iter()
-            .flat_map(|(epoch_slots, stake)| {
-                epoch_slots
-                    .to_slots(root)
-                    .filter(|slot| slot_range.contains(slot))
-                    .zip(std::iter::repeat((epoch_slots.from, *stake)))
-            })
-            .into_group_map();
-        let slot_nodes_stakes: Vec<_> = {
-            let mut cluster_slots = self.cluster_slots.write().unwrap();
-            slot_nodes_stakes
-                .into_iter()
-                .map(|(slot, nodes_stakes)| {
-                    let slot_nodes = cluster_slots.entry(slot).or_default().clone();
-                    (slot_nodes, nodes_stakes)
-                })
-                .collect()
         };
-        for (slot_nodes, nodes_stakes) in slot_nodes_stakes {
-            slot_nodes.write().unwrap().extend(nodes_stakes);
+        // a handy closure to spawn new hashmaps with the correct parameters
+        let map_maker = || {
+            let mut map =
+                HashMap::with_capacity_and_hasher(map_capacity, PubkeyHasherBuilder::default());
+            zero_init_map(&mut map);
+            self.metric_allocations.fetch_add(1, Ordering::Relaxed);
```

Comment:
nice, this should be an interesting metric 😃

Reply:
It is stuck at zero, most of the time, not very interesting. Mostly this is a failsafe in case of regressions.